gangplank/ocp/bc.go

```diff
@@ -223,136 +214,277 @@ binary build interface.`)
-						&RemoteFile{
-							Bucket: "builds",
-							Minio:  m,
-							Object: keyPath,
-						},
-					)
-					log.WithField("file", keyPath).Info("Sending meta-data to worker")
-					return nil
-				})
-
-				for _, artifact := range s.RequireArtifacts {
-					bArtifact, err := mBuild.GetArtifact(artifact)
-					if err != nil {
-						log.WithFields(log.Fields{
-							"stage":    s.ID,
-							"artifact": artifact,
-						}).Info("Waiting on artifact")
-						time.Sleep(15 * time.Second)
-						break
+
+					// Once all the dependencies are met, inform the out channel exit.
+					if len(s.RequireArtifacts) == foundCount {
+						out <- true
+						return true
 					}
+					return false
+				}
 
-					// get the Minio relative path for the object
-					// the full path needs to be broken in to <BUILDID>/<ARCH>/<FILE>
-					keyPath := filepath.Join(buildPath, filepath.Base(bArtifact.Path))
-
-					log.WithFields(log.Fields{
-						"artifact path": keyPath,
-						"artifact":      artifact,
-						"buildID":       buildID,
-						"stage":         s.ID,
-					}).Info("Found required artifact")
-
-					r := &RemoteFile{
-						Artifact: bArtifact,
-						Bucket:   "builds",
-						Minio:    m,
-						Object:   keyPath,
+				for {
+					if check() {
+						l.Debug("all dependencies for stage have been meet")
+						break
+					}
+					// Wait for the next check or terminate.
+					select {
+					case <-terminate:
+						return
+					case <-time.After(15 * time.Second):
+						return
 					}
-					ws.RemoteFiles = append(ws.RemoteFiles, r)
+				}
+			}(out)
+			return out
+		}
+
+		// anonFunc performs the actual work..
+		anonFunc := func(wg *sync.WaitGroup, terminate <-chan bool, errCh chan<- error) {
+			defer wg.Done()
+
+			ws := &workSpec{
+				APIBuild:      apiBuild,
+				ExecuteStages: []string{s.ID},
+				JobSpec:       bc.JobSpec,
+				RemoteFiles:   remoteFiles,
+				Return:        returnTo,
+			}
+
+			select {
+			case <-terminate:
+				l.Warning("Terminate signal recieved, aborting stage")
+				return
+			case <-time.After(60 * time.Minute):
+				l.Error("Timed out waiting for dependencies")
+				errCh <- errors.New("required artifacts never appeared")
+				return
+			case ok := <-ready(ws, terminate):
+				if !ok {
+					errCh <- fmt.Errorf("%s failed to become ready", s.ID)
+					return
+				}
+
+				l.Info("Worker dependences have been defined")
+				eVars, err := ws.getEnvVars()
+				if err != nil {
+					errCh <- err
+					return
 				}
 
-				log.WithField("stage", s.ID).Infof("All dependencies for stage have been meet")
-				ready = true
+				l.Info("Executing worker pod")
+				if err := cpod.WorkerRunner(podCtx, eVars); err != nil {
+					l.WithError(err).Error("Failed stage execution")
+					err = fmt.Errorf("%s failed: %w", s.ID, err)
+					errCh <- err
+					return
+				}
 			}
 		}
 
-		eVars, err := ws.getEnvVars()
-		if err != nil {
-			return err
+		// If there is no default execution order, default to 2
```

Comment:
What's the rationale for this?

Reply:
I updated the comment to answer the question: basically, some stage short-hands have automatic stage assignment (`cosa build` aka `base` is 1, `cosa meta --finalize` aka `finalize` is 999). A follow on PR might to add execution order numbers if none is provided which might make the thinking a bit more clear here.